Benchmark test for Kuka arm inverse dynamics

mitiguy opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7214)
<!-- Reviewable:end -->

mitiguy commented:
Feature review +@amcastro-tri  
Platform review +@seancurtis-tri 

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-Kvou-9vUvLfG3zukuYK:b-50tqzl)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxaR5Fq9QvI6I6BWIn:-KvxaR5Fq9QvI6I6BWIo:b1pe4jf) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L176)):*
> ```objc
> 
>     // Add force element for a constant gravity pointing downwards, that is, in
>     // the minus y-axis direction.
> ```

the comment says `y-axis` but the code is in the `z-axis`.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxaWb2yHBOZhF-xCmI:-KvxaWb2yHBOZhF-xCmJ:bba89or) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L179)):*
> ```objc
>     set_gravity(gravity);
>     const Eigen::Vector3d gravity_vector = gravity_ * Eigen::Vector3d::UnitZ();
>     model_->AddForceElement<UniformGravityFieldElement>(gravity_vector);
> ```

Notice that `UniformGravityFieldElement` does exactly what you ask for. That is, here you are asking gravity to point in the positive z-axis direction. Is that what you want?
Maybe you'd need to change to `const Eigen::Vector3d gravity_vector = -gravity_ * Eigen::Vector3d::UnitZ();` (notice minus sign).

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxaBY0eqE_hgZ2CTro:-KvxaBY0eqE_hgZ2CTrp:b4wpdle) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h#L59)):*
> ```objc
>   /// model of the aforementioned KUKA robot.  All model parameters are from:
>   /// drake/multibody/benchmarks/kuka_iiwa_robot/kuka_iiwa_robot.urdf
>   explicit MGKukaIIwaRobot(double gravity) {
> ```

could you comment in the constructor in what direction gravity points? `gravity > 0` implies the gravity vector being `-gravity * ẑ`?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kvxa4CJ0dJcx0KX1fNp:-Kvxa4CJ0dJcx0KX1fNq:bqytut5) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h#L180)):*
> ```objc
>   /// @param[in] gravity Earth's gravitational acceleration in m/s².
>   void set_gravity(double gravity) { MG_kuka_auto_generated_.g = gravity; }
> ```

BTW, so `gravity > 0` implies the gravity vector pointing in the minus z-axis direction?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxavKTvydGDKeS5bNb:-KvxavKU-gQTE7LxzRKa:b-w9ka1g) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L108)):*
> ```cpp
>   // TODO(@mitiguy) Check why setting gravityB = 9.8 does not work.
>   const double gravityA = 0;
>   const double gravityB = 0*9.8;
> ```

I believe there might be a problem with the sign of the gravity vector in the MBT model. See comment above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxaR5Fq9QvI6I6BWIn:-Kw2VgM6XIYxcbbXmxCK:b-6ztgbi) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L176)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

the comment says `y-axis` but the code is in the `z-axis`.
</blockquote></details>

Done.  Comment fixed.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxaWb2yHBOZhF-xCmI:-Kw2VqCNLtqLCwUN8SDO:b-p1izqe) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L179)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Notice that `UniformGravityFieldElement` does exactly what you ask for. That is, here you are asking gravity to point in the positive z-axis direction. Is that what you want?
Maybe you'd need to change to `const Eigen::Vector3d gravity_vector = -gravity_ * Eigen::Vector3d::UnitZ();` (notice minus sign).
</blockquote></details>

Done.  Agreed - added negative sign.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxaBY0eqE_hgZ2CTro:-Kw2X9C60PXH3Ql0iJf1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h#L59)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

could you comment in the constructor in what direction gravity points? `gravity > 0` implies the gravity vector being `-gravity * ẑ`?
</blockquote></details>

Done.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxavKTvydGDKeS5bNb:-Kw2ZPLs_Mw-sCxRCU7G:b-eoevzr) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L108)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I believe there might be a problem with the sign of the gravity vector in the MBT model. See comment above.
</blockquote></details>

Done.  I agree with your comment and all seems well now.
Thanks for finding the negative sign error.  
Note: Late Friday night, the benchmark worked well,
except for the gravity, hence the TODO to myself when
I could revisit the code with a fresh mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you so much for the test Paul! Just minor comments. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 5 of 9 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/BUILD, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG1e2JrcCtg1uphsP1:-KwG1e2JrcCtg1uphsP2:bvj6q2m) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/BUILD#L22)):*
> ```Python
>     ],
> )
> ```

BTW, you'll need to merge with master and rename this file to `BUILD.bazel`, per #7237 

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 57 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG0kSbK90xut6tB40h:-KwG0kSbK90xut6tB40i:b-rxc5xi) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L57)):*
> ```objc
> /// a_NBo_N    | Bo's acceleration in N, expressed in N.
> template <typename T>
> struct RigidBodyKinematics {
> ```

This name might go into conflict with other future parts of the code. To keep it consistent with `KukaRobotJointReactionForces`, which includes the prefix `KukaRobot`, why not to name this struct something lik `KukaRobotKinamtics` instead?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 291 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG125z2ls_Y7oN9YVU:-KwG125z2ls_Y7oN9YVV:b-89bsj2) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L291)):*
> ```objc
>     // model_->AddForceElement<UniformGravityFieldElement>(gravity_vector);
>     model_->CalcForceElementsContribution(*context_, pc, vc,
>                           &Fapplied_Bo_W_array, &generalized_force_applied);
> ```

BTW, not sure but, odd indentation?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 465 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG1U-vZjm4M9ZxBA0v:-KwG1U-vZjm4M9ZxBA0w:bd65guy) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L465)):*
> ```objc
>   // After model is finalized, create default context.
>   std::unique_ptr<systems::Context<double>> context_;
>   MultibodyTreeContext<double>* mbt_context_;
> ```

Not used (besides being initialized).

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KvxavKTvydGDKeS5bNb:-KwFzjtuEc0C43XurFyE:blsocdy) ([raw file](https://github.com/robotlocomotion/drake/blob/9394d320b7d9b79758507414d1f9e359eb42d488/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L108)):*
<details><summary><i>Previously, mitiguy wrote…</i></summary><blockquote>

Done.  I agree with your comment and all seems well now.
Thanks for finding the negative sign error.  
Note: Late Friday night, the benchmark worked well,
except for the gravity, hence the TODO to myself when
I could revisit the code with a fresh mind.
</blockquote></details>

The good news is that now this code verifies MBT implementation for a true 3D case! thanks!

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwFzzYEC55wpBsj975T:-KwFzzYEC55wpBsj975U:b-ufh2hq) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L4)):*
> ```cpp
> 
> #include "drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h"
> #include "drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h"
> ```

BTW order of includes

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG-74J-pULbk2n-LvB:-KwG-74J-pULbk2n-LvC:b-e1di7) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L19)):*
> ```cpp
> 
> // Compare Drake's MultibodyTree forces and motion with MotionGenesis solution.
> void TestKukaArmInverseDynamics(const Eigen::Ref<const VectorX<double>> &q,
> ```

BTW & must go next to type.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG-gBFTPGHLDYK6Ga7:-KwG-gBFTPGHLDYK6Ga8:b-blx5iu) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L72)):*
> ```cpp
>   // Inverse dynamics: Get corresponding MotionGenesis information.
>   SpatialForced F_Ao_W, F_Bo_W, F_Co_W, F_Do_W, F_Eo_W, F_Fo_W, F_Go_W;
>   std::tie(F_Ao_W, F_Bo_W, F_Co_W, F_Do_W, F_Eo_W, F_Fo_W, F_Go_W) =
> ```

BTW. Interesting you decided the MG and MBT classes to have different API's with different return types. Any reason for that? I think it'd read more nicely if they both had the same API's. Up to you, this works.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 27 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG0Q-T7Z85rGQgdZss:-KwG0Q-T7Z85rGQgdZst:b-rxc5xi) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L27)):*
> ```cpp
>   // a_NGo_N    | Go's acceleration in N, expressed in N.
>   DrakeKukaIIwaRobot drake_kuka_robot(0);
>   const RigidBodyKinematics<double> kinematics =
> ```

This name might go into conflict with other future parts of the code. To keep it consistent with `KukaRobotJointReactionForces`, which includes the prefix `KukaRobot`, why not to name this struct something lik `KukaRobotKinamtics` instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-KwG25YVrYbX3iNYbezu:b-8g9xt)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
Reviewed 1 of 9 files at r1, 3 of 4 files at r2.
Review status: 5 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 57 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG0kSbK90xut6tB40h:-KwNX1ztOwkMc8Cu2pGH:b-ccalkt) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L57)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This name might go into conflict with other future parts of the code. To keep it consistent with `KukaRobotJointReactionForces`, which includes the prefix `KukaRobot`, why not to name this struct something lik `KukaRobotKinamtics` instead?
</blockquote></details>

Done.  Pushed this to multibody_tree/test_utilities.
New classes there are:
RigidBodyKinematicsPVA
SpatialKinematicsPVA.

I regard one or both of these classes as having general utility
(I will use them ofr testing).  Sherm suggests that until they 
are needed more often, to leave the sugar in the new
folder multibody_tree/test_utilities.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 465 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG1U-vZjm4M9ZxBA0v:-KwNYTLdB-cyHlvQ63sT:btesnx1) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L465)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Not used (besides being initialized).
</blockquote></details>

Done.  Removed.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwFzzYEC55wpBsj975T:-KwN_IC5A3TkZ2QbVzp5:b-nd5juv) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L4)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW order of includes
</blockquote></details>

I tried to switch them and cpplint complained (could not complete bazel ... test ).
So I switched them back.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 27 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG0Q-T7Z85rGQgdZss:-KwNZX-9o0Uxi1OrjMob:brot162) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L27)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This name might go into conflict with other future parts of the code. To keep it consistent with `KukaRobotJointReactionForces`, which includes the prefix `KukaRobot`, why not to name this struct something lik `KukaRobotKinamtics` instead?
</blockquote></details>

Done.  See previous comment and new changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
@ drake-junkins-bot  linux-xenial-clang-bazel-experimental-debug   please

mitiguy commented:
@drake-jenkins-bot  linux-xenial-clang-bazel-experimental-everything  please

mitiguy commented:
@drake-jenkins-bot  linux-xenial-gcc-bazel-experimental  please

mitiguy commented:
@drake-jenkins-bot  linux-xenial-clang-bazel-experimental-debug  please

jwnimmer-tri commented:
@mitiguy Apologies, but this PR has been bitten by issue #7271, so you are not getting good CI feedback.  To fix this PR's lint, you should run `bazel-bin/tools/lint/buildifier drake/multibody/multibody_tree/test_utilities/BUILD.bazel`, and then commit its suggestions.

amcastro-tri commented:
Reviewed 6 of 8 files at r3.
Review status: 11 of 12 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 57 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG0kSbK90xut6tB40h:-Kwjdi581iwFEWfHmW1c:b8e3zvy) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L57)):*
<details><summary><i>Previously, mitiguy wrote…</i></summary><blockquote>

Done.  Pushed this to multibody_tree/test_utilities.
New classes there are:
RigidBodyKinematicsPVA
SpatialKinematicsPVA.

I regard one or both of these classes as having general utility
(I will use them ofr testing).  Sherm suggests that until they 
are needed more often, to leave the sugar in the new
folder multibody_tree/test_utilities.
</blockquote></details>

BTW This is not what I originally was talking about in the comment. There was no need to move or add more code. In any case, I am happy with any other solution you might come with. Just for the record this is clearly not what I originally asked for.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjg69ddwaqCOz91RsR:-Kwjg69ddwaqCOz91RsS:b-rms7j5) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L1)):*
> ```objc
> #pragma once
> ```

BTW I just realize we decided to place this code (the one under test) within the `benchmarks` directory. I can see  how for instance MG code would live in `benchmarks` since that is the actual "benchmarking code". Now, MBT is the code under test, shouldn't then `drake_kuka_iiwa_robot.h` live in `multibody_tree/test/kukka_iiwa_robot/`? @sherm1? what do you think?
For instance, the `MG` code in `benchmarks` can potentially be used to verify different multibody implementations (we are ok with one of course), but following this logic you can see how the babyboot  can be used to test some of Evan's contact code as well as to test some MBT code.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 15 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjdWGBoMF60hmKhkiL:-KwjdWGBoMF60hmKhkiM:bx68gnh) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L15)):*
> ```objc
> #include "drake/multibody/multibody_tree/revolute_mobilizer.h"
> #include "drake/multibody/multibody_tree/rigid_body.h"
> #include "drake/multibody/multibody_tree/test_utilities/rigid_body_kinematics_PVA.h"
> ```

naming violates the style guide. Just name `spatial_kinematics.h`?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 188 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjeeXS3FY9mQYka2BQ:-KwjeeXS3FY9mQYka2BR:brgaewj) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L188)):*
> ```objc
>   ///
>   /// @returns G's kinematics in N, expressed in N.
>   SpatialKinematicsPVA<double>
> ```

BTW I actually like how this reads. But to avoid violating the style guide I'd rename to just `SpatialKinematics`

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 237 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjfJUGtceuja13T2QY:-KwjfJUGtceuja13T2QZ:bbiwgxe) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L237)):*
> ```objc
>   /// F_Go_W  | Spatial force on Go from F, expressed in frame W (world).
>   const KukaRobotJointReactionForces
>       CalcJointReactionForces(const Eigen::Ref<const VectorX<double>>& q,
> ```

BTW odd indentation

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjh8xWqut4kGI4RDpN:-Kwjh8xWqut4kGI4RDpO:bl9rux) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/MG/MG_kuka_iiwa_robot.h#L86)):*
> ```objc
>   /// alpha_NG_N | G's angular acceleration in N, expressed in N.
>   /// a_NGo_N    | Go's acceleration in N, expressed in N.
>   std::tuple<Matrix3d, Vector3d, Vector3d, Vector3d, Vector3d, Vector3d>
> ```

BTW we could here no problem to return a `SpatialKinematics` if you wanted to, for consistency with Drake's MBT counterpart.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwFzzYEC55wpBsj975T:-Kwje9kgO1gSzYim69CI:bkeezyj) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L4)):*
<details><summary><i>Previously, mitiguy wrote…</i></summary><blockquote>

I tried to switch them and cpplint complained (could not complete bazel ... test ).
So I switched them back.
</blockquote></details>

BTW I think that might have to do with Jeremy's comment?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwji1tee_9fEvlpUB_x:-Kwji1tee_9fEvlpUB_y:bx41ohc) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L38)):*
> ```cpp
> 
>   // Kinematics: Get corresponding MotionGenesis information.
>   // TODO(Mitiguy) Have MotionGenesis also return RigidBodyKinematicsPVA.
> ```

I think the right object to return is `SpatialKinematics`. I see no need for class `RigidBodyKinematics`. 

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 27 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwG0Q-T7Z85rGQgdZss:-KwjeG3oUEOFxsAIjhPd:b-gvxr0e) ([raw file](https://github.com/robotlocomotion/drake/blob/6effd9a203db79715bbe465f8b98f9ce6b79f942/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L27)):*
<details><summary><i>Previously, mitiguy wrote…</i></summary><blockquote>

Done.  See previous comment and new changes.
</blockquote></details>

BTW. Same answer. The new changes introduced  clearly go beyond the simple point I was trying to make (a name change only).

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhNsU36HsXtFzvInj:-KwjhNsU36HsXtFzvInk:b6y2nl9) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L31)):*
> ```cpp
>   const SpatialKinematicsPVA<double> spatial_kinematics =
>       drake_kuka_robot.CalcEndEffectorKinematics(q, qDt, qDDt);
>   const RigidBodyKinematicsPVA<double> kinematics(spatial_kinematics);
> ```

This in my mind just generates two classes instead of one (`SpatialKinematics`) where you could have appropriate getters to get this information.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhZeCly5R1bwIjYme:-KwjhZeCly5R1bwIjYmf:b-hqgfw5) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L44)):*
> ```cpp
>   // Compare actual results with expected (true) results.
>   constexpr double kEpsilon = 10 * std::numeric_limits<double>::epsilon();
>   EXPECT_TRUE(kinematics.R_NB_.isApprox(R_NG_true, kEpsilon));
> ```

This could read `spatial_kinematics.get_orientation()`

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 46 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhdmZwFJMz96mBnKL:-KwjhdmZwFJMz96mBnKM:b-3m7qjk) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L46)):*
> ```cpp
>   EXPECT_TRUE(kinematics.R_NB_.isApprox(R_NG_true, kEpsilon));
>   EXPECT_TRUE(kinematics.p_NoBo_N_.isApprox(p_NoGo_N_true, kEpsilon));
>   EXPECT_TRUE(kinematics.w_NB_N_.isApprox(w_NG_N_true, kEpsilon));
> ```

This could read This could read `spatial_kinematics.get_angular_velocity()`

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhhtP0P4WARpEZio7:-KwjhhtP0P4WARpEZio8:b6kh6fe) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L49)):*
> ```cpp
>   EXPECT_TRUE(kinematics.v_NBo_N_.isApprox(v_NGo_N_true, kEpsilon));
>   EXPECT_TRUE(kinematics.alpha_NB_N_.isApprox(alpha_NG_N_true, kEpsilon));
>   EXPECT_TRUE(kinematics.a_NBo_N_.isApprox(a_NGo_N_true, kEpsilon));
> ```

This could read This could read `spatial_kinematics.get_linear_acceleration()`

---

*[drake/multibody/multibody_tree/test_utilities/rigid_body_kinematics_PVA.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjc0xJTLe2wwswKcuo:-Kwjc0xJTLe2wwswKcup:b-2dd7u8) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/rigid_body_kinematics_PVA.h#L26)):*
> ```objc
> /// a_NBo_N    | Bo's acceleration in N, expressed in N.
> template <typename T>
> class RigidBodyKinematicsPVA {
> ```

This seems unnecessarily repeating code from `SpatialKinematics`. Why not to just add getters to `SpatialKinematics`? like `Vector3<T> SpatialKinematics::get_position() const`, `Matrix3<T> SpatialKinematics::get_orientation() const` and so on.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 11 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjdKHqILdvkYelbmIP:-KwjdKHqILdvkYelbmIQ:bfuhbms) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L11)):*
> ```objc
> namespace multibody_tree {
> namespace test_utilities {
> ```

BTW, I would move this class somewhere up the chain so that both `benchmarks` and `multibody` code can use it. So that you get similar API's!

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 13 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjaQBPScVE9Z1604Vm:-KwjaQBPScVE9Z1604Vn:b-fyx77o) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L13)):*
> ```objc
> 
> /// Utility class containing the spatial position, velocity, and acceleration
> /// of a rigid frame B in another arbitrary rigid frame N (e.g., the world).
> ```

BTW we do not have the concept of a "rigid frame" (and probably we will never have it) in Drake. Just say "frame"?

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjcqaO5z8eO3t6h8qX:-KwjcqaO5z8eO3t6h8qY:bc1lspk) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L26)):*
> ```objc
> ///         | in N).
> template <typename T>
> class SpatialKinematicsPVA {
> ```

Border line violating the style guide. I would just call this `SpatialKinematics`. You did a great job documenting what this class has in it so no need to pollute its name with more details (in a from that actually breaks our conventions).

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjbf-o8k6o3lHaP22R:-Kwjbf-o8k6o3lHaP22S:b-tfzph2) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L35)):*
> ```objc
>   SpatialKinematicsPVA(const Isometry3<T>& X_NB,
>                        const SpatialVelocity<T>& V_NBo_N,
>                        const SpatialAcceleration<T>& A_NBo_N) {
> ```

BTW use initializer list?

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 41 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjceko7q2ALLVYs2hM:-Kwjceko7q2ALLVYs2hN:be8noge) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L41)):*
> ```objc
>   }
> 
>   Isometry3<T> X_NB_;
> ```

`private:`? and provide getters?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
A few comments ...

---

Review status: 11 of 12 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjg69ddwaqCOz91RsR:-KwlILz0TB00u6tkYrEE:bwtvfqo) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L1)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW I just realize we decided to place this code (the one under test) within the `benchmarks` directory. I can see  how for instance MG code would live in `benchmarks` since that is the actual "benchmarking code". Now, MBT is the code under test, shouldn't then `drake_kuka_iiwa_robot.h` live in `multibody_tree/test/kukka_iiwa_robot/`? @sherm1? what do you think?
For instance, the `MG` code in `benchmarks` can potentially be used to verify different multibody implementations (we are ok with one of course), but following this logic you can see how the babyboot  can be used to test some of Evan's contact code as well as to test some MBT code.
</blockquote></details>

Yeah, that makes sense @amcastro-tri. It would be nice to have unit tests as close as possible to the code they are testing. And we would like to have the reusable benchmark code under benchmarks. Should we attempt that in this PR or merge it with the current structure and an issue saying it should be refactored when we have a second use for it?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhZeCly5R1bwIjYme:-KwlJIu0zAr88JIwWpNI:bqq0rpg) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L44)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This could read `spatial_kinematics.get_orientation()`
</blockquote></details>

BTW consider just `spatial_kinematics.orientation()` -- we've been moving towards dropping the `get_` for these very straightforward accessors. (The pair would then be `orientation()` and `set_orientation()`.) This is allowed by the styleguide and very commonly used (consider `vector.size()`, even `isometry.translation()`).

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 13 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjaQBPScVE9Z1604Vm:-KwlKIsUb_nWZQxeqDHb:bzf301q) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L13)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW we do not have the concept of a "rigid frame" (and probably we will never have it) in Drake. Just say "frame"?
</blockquote></details>

BTW I agree -- even the frames we have in mind for flexible bodies will be rigid, in the sense that their axes remain orthogonal.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjbf-o8k6o3lHaP22R:-KwlKX7m87YC1l9A25T9:b-powsld) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L35)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW use initializer list?
</blockquote></details>

BTW initializer lists require same-type elements (like an std::vector).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-KwlKr_f0VH24D-vuHm8:blymdu6)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
Review status: 2 of 12 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 15 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjdWGBoMF60hmKhkiL:-KwlGZE4IEwZHBKKgwG_:b3w8qor) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L15)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

naming violates the style guide. Just name `spatial_kinematics.h`?
</blockquote></details>

I envision needing to also have a class named SpatialKinematicsPV 
(in addition to SpatialKinematicsPVA).

Both these names are specific to our purposes and eary to understand.
I see (below) that you like this name - and hopefully, you agree to the
logic as why it is named as such.  I prefer to leave the name as is.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwji1tee_9fEvlpUB_x:-KwlIjLbd08sT4SvmtJm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L38)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think the right object to return is `SpatialKinematics`. I see no need for class `RigidBodyKinematics`. 
</blockquote></details>

Done.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhNsU36HsXtFzvInj:-KwmDj2pG0_YJSPdtQoE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L31)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This in my mind just generates two classes instead of one (`SpatialKinematics`) where you could have appropriate getters to get this information.
</blockquote></details>

Done.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhZeCly5R1bwIjYme:-KwmDnaNhRKZJ5wo0YCR:b-naljgh) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L44)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider just `spatial_kinematics.orientation()` -- we've been moving towards dropping the `get_` for these very straightforward accessors. (The pair would then be `orientation()` and `set_orientation()`.) This is allowed by the styleguide and very commonly used (consider `vector.size()`, even `isometry.translation()`).
</blockquote></details>

Since there are various representations of orientation, I prefer
get_rotation_matrix(), -- which is now been shortened to simply rotation_matrix().

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 46 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhdmZwFJMz96mBnKL:-KwmEQFRUTB-kr1KYY7T:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This could read This could read `spatial_kinematics.get_angular_velocity()`
</blockquote></details>

Done.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjhhtP0P4WARpEZio7:-KwmERJMhZBG54fAGx1k:bc7u1t8) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_kinematics_test.cc#L49)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This could read This could read `spatial_kinematics.get_linear_acceleration()`
</blockquote></details>

Done.  As in Alejandro's SpatialAcceleration class, now uses translational_acceleration().

---

*[drake/multibody/multibody_tree/test_utilities/rigid_body_kinematics_PVA.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjc0xJTLe2wwswKcuo:-KwmEarE4TGwH4nw3QCi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/rigid_body_kinematics_PVA.h#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This seems unnecessarily repeating code from `SpatialKinematics`. Why not to just add getters to `SpatialKinematics`? like `Vector3<T> SpatialKinematics::get_position() const`, `Matrix3<T> SpatialKinematics::get_orientation() const` and so on.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 11 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjdKHqILdvkYelbmIP:-KwmGSRRnIzZiPrNmylG:bwj0y16) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L11)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, I would move this class somewhere up the chain so that both `benchmarks` and `multibody` code can use it. So that you get similar API's!
</blockquote></details>

I agree.  We can talk about this when you are here.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KwjcqaO5z8eO3t6h8qX:-KwmG6Gg051zp4B_cdIu:b564ue0) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Border line violating the style guide. I would just call this `SpatialKinematics`. You did a great job documenting what this class has in it so no need to pollute its name with more details (in a from that actually breaks our conventions).
</blockquote></details>

As mentioned earlier,  I envision needing to 
also have a class named SpatialKinematicsPV 
(in addition to SpatialKinematicsPVA).

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h, line 41 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjceko7q2ALLVYs2hM:-KwmFfr1on7YO1aix94h:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/multibody_tree/test_utilities/spatial_kinematics_PVA.h#L41)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`private:`? and provide getters?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 2 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwjg69ddwaqCOz91RsR:-KwvhHWVJxoRpw23ABSH:b-pmtcy) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L1)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yeah, that makes sense @amcastro-tri. It would be nice to have unit tests as close as possible to the code they are testing. And we would like to have the reusable benchmark code under benchmarks. Should we attempt that in this PR or merge it with the current structure and an issue saying it should be refactored when we have a second use for it?
</blockquote></details>

(Discussed with Paul.) Now I'm not sure it would make sense to move this header out of benchmarks because it is critical that it precisely match (a) the MG code, and (b) the urdf that is kept here for reference. And we never intended that benchmarks/ would hold only externally-generate stuff; it is supposed to contain lots of Drake code also. But I do agree that specific unit tests that *employ* these benchmark models would be better placed in the test/ directory of the code under test. Probably not in this PR though -- unblocking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 11 of 12 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/BUILD.bazel, line 20 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxAIsfDyX1qHifHd5mn:-KxAIsfDyX1qHifHd5mo:b-5fnafd) ([raw file](https://github.com/robotlocomotion/drake/blob/7c4b400d882bfa9e49363a367209ba27d175c35f/drake/multibody/benchmarks/kuka_iiwa_robot/BUILD.bazel#L20)):*
> ```
>         "//drake/multibody/multibody_tree",
>         "//drake/multibody/multibody_tree/test_utilities:spatial_kinematics",  # noqa
>     ],
> ```

BTW is this over 80 characters now?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/MG/BUILD.bazel, line 26 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxAJ4q79xciTok3ptu3:-KxAJ4q79xciTok3ptu4:b6zgb35) ([raw file](https://github.com/robotlocomotion/drake/blob/7c4b400d882bfa9e49363a367209ba27d175c35f/drake/multibody/benchmarks/kuka_iiwa_robot/MG/BUILD.bazel#L26)):*
> ```
>         "//drake/math:geometric_transform",
>         "//drake/multibody/multibody_tree",
>         "//drake/multibody/multibody_tree/test_utilities:spatial_kinematics",  # noqa
> ```

BTW, 80 characters long?

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-Kwji1tee_9fEvlpUB_x:-KxAKChAVvc0Ps2KT3ka:bpwxhpt) ([raw file](https://github.com/robotlocomotion/drake/blob/83765df383e4919b286b2a3e0574b9122474aafa/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L38)):*
<details><summary><i>Previously, mitiguy wrote…</i></summary><blockquote>

Done.
</blockquote></details>

BTW TODO not needed anymore.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc, line 18 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxAK3gNw_JHIs96LyYI:-KxAK3gNw_JHIs96LyYJ:b-7uk3pi) ([raw file](https://github.com/robotlocomotion/drake/blob/7c4b400d882bfa9e49363a367209ba27d175c35f/drake/multibody/benchmarks/kuka_iiwa_robot/test/kuka_iiwa_robot_inverse_dynamics_test.cc#L18)):*
> ```cpp
> using Vector7d = Eigen::Matrix<double, 7, 1>;
> ```

BTW blank space

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h, line 29 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxAGSgaZJNBtHOu750A:-KxAGSgaZJNBtHOu750B:b6ybnnv) ([raw file](https://github.com/robotlocomotion/drake/blob/7c4b400d882bfa9e49363a367209ba27d175c35f/drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h#L29)):*
> ```objc
>                        const SpatialVelocity<T>& V_NBo_N,
>                        const SpatialAcceleration<T>& A_NBo_N) :
>                        X_NB_(X_NB), V_NBo_N_(V_NBo_N), A_NBo_N_(A_NBo_N) {;}
> ```

BTW colon not needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
-@SeanCurtis-TRI  +@sammy-tri  For platform review, please per rotation.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-KxAxP5Kw46BXYDjs4Ul:b-16ftwo)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+(status: curate commits before merging) I'll be blocked in meetings for a while yet, but I should be able to get a first pass done today.

---

Reviewed 1 of 9 files at r1, 1 of 12 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-KxEVttZ6jn3oiHpH6No:bkgt7v6)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass complete, just a couple of minor comments.

---

Reviewed 9 of 12 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 376 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEnePCQ4fzPTzICMP5:-KxEnePD2DT-DLVFoL2W:b4i50ek) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L376)):*
> ```objc
>   // are fixed in N (Earth) with Nz vertically upward (so gravity is in -Nz).
>   // @param[in] gravity Earth's gravitational acceleration in m/s².
>   void set_gravity(double gravity) {gravity_ = gravity;}
> ```

http://drake.mit.edu/styleguide/cppguide.html#Horizontal_Whitespace says "For inline function implementations, put spaces between the braces and the implementation itself."  (this also applies to `get_number_of_rigid_bodies()`)

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h, line 29 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEoKysRjRwyqgPxn2v:-KxEoKysRjRwyqgPxn2w:b-xxfzll) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h#L29)):*
> ```objc
>                        const SpatialVelocity<T>& V_NBo_N,
>                        const SpatialAcceleration<T>& A_NBo_N) :
>                        X_NB_(X_NB), V_NBo_N_(V_NBo_N), A_NBo_N_(A_NBo_N) {;}
> ```

Superfluous ';'

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h, line 114 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEp4kJRD-amItwBPMk:-KxEp4kJRD-amItwBPMl:b-t6hqzp) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h#L114)):*
> ```objc
>   // To avoid changing the interface to this class, additional member data is
>   // added to class to compensate for the Isometry3 member methods linear() and
>   // translation() which do not return references.
> ```

I'm a little unclear on how returning a reference when you're doing a value assignment under the hood (to populate the thing you're returning the const reference to) anyway is superior to returning by value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-KxEqfoTjevnhiyILTUJ:b-k3p5ly)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
Review status: 7 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 376 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEnePCQ4fzPTzICMP5:-KxJYnayFZ-75c9KCnzt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L376)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

http://drake.mit.edu/styleguide/cppguide.html#Horizontal_Whitespace says "For inline function implementations, put spaces between the braces and the implementation itself."  (this also applies to `get_number_of_rigid_bodies()`)
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h, line 29 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEoKysRjRwyqgPxn2v:-KxJaPQLVVbIcDiCAWtB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h#L29)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Superfluous ';'
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h, line 114 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEp4kJRD-amItwBPMk:-KxJZKaQMfGF7fMPhp_i:bed4cgo) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h#L114)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'm a little unclear on how returning a reference when you're doing a value assignment under the hood (to populate the thing you're returning the const reference to) anyway is superior to returning by value.
</blockquote></details>

The goal is to make the interface to this class consistent.
Four of the six returned quantities are returned by reference.
The last two dig information out of an Isometry3 and cannot return by reference.
The good news is that the Isometry3 class will soon be superseded by the Transform class.
The TODO(Mitiguy) is a reminder to upgrade to the new Transform class when ready.
However, I do not want the interface to change when this occurs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h, line 114 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-KxEp4kJRD-amItwBPMk:-KxKQ2ed8Ji7ZFcohSVA:b-hg7hgb) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef7906a55916b09f659c7341953ac60257893d1/drake/multibody/multibody_tree/test_utilities/spatial_kinematics.h#L114)):*
<details><summary><i>Previously, mitiguy wrote…</i></summary><blockquote>

The goal is to make the interface to this class consistent.
Four of the six returned quantities are returned by reference.
The last two dig information out of an Isometry3 and cannot return by reference.
The good news is that the Isometry3 class will soon be superseded by the Transform class.
The TODO(Mitiguy) is a reminder to upgrade to the new Transform class when ready.
However, I do not want the interface to change when this occurs.
</blockquote></details>

Ok, I guess if these go away when the TODO is resolved I'm alright with it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7214#-:-KxKQOTncFSAB4v1azUn:bnfp4nl)*
<!-- Sent from Reviewable.io -->
